diff --git a/llvm/include/llvm/CodeGen/TargetInstrInfo.h b/llvm/include/llvm/CodeGen/TargetInstrInfo.h
index 72f69f4c6..3d109a6d9 100644
--- a/llvm/include/llvm/CodeGen/TargetInstrInfo.h
+++ b/llvm/include/llvm/CodeGen/TargetInstrInfo.h
@@ -1208,6 +1208,8 @@ public:
   /// even if it has glue.
   virtual bool canCopyGluedNodeDuringSchedule(SDNode *N) const { return false; }
 
+  virtual int getJumpTableIndex(const MachineInstr &MI) const { return -1; }
+
 protected:
   /// Target-dependent implementation for foldMemoryOperand.
   /// Target-independent code in foldMemoryOperand will
diff --git a/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp b/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp
index 32a10ad41..7701cac65 100644
--- a/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp
+++ b/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp
@@ -125,6 +125,12 @@

 using namespace llvm;

+
+static cl::opt<bool>
+    AddIndirectBranchDestAnnotationForDeegen("add-indirect-branch-dest-annotation-for-deegen", cl::init(false), cl::Hidden,
+                                            cl::desc("Add destination annotations to indirect branches for Deegen"));
+
+
 #define DEBUG_TYPE "asm-printer"

 const char DWARFGroupName[] = "dwarf";
@@ -1490,6 +1496,35 @@ void AsmPrinter::emitFunctionBody() {
       if (isVerbose())
         emitComments(MI, OutStreamer->getCommentOS());

+      if (AddIndirectBranchDestAnnotationForDeegen)
+      {
+        const TargetInstrInfo* TII = MF->getSubtarget().getInstrInfo();
+        if (TII->isTailCall(MI))
+        {
+            OutStreamer->AddComment("__deegen_asm_annotation_tailcall");
+        }
+        else
+        {
+          int jumpTableIndex = TII->getJumpTableIndex(MI);
+          if (jumpTableIndex != -1)
+          {
+            const std::vector<MachineJumpTableEntry>& jumpTables = MF->getJumpTableInfo()->getJumpTables();
+            if (jumpTableIndex < 0 || static_cast<size_t>(jumpTableIndex) >= jumpTables.size())
+            {
+                fprintf(stderr, "Deegen AsmPrinter hack: Bad jump table index!\n");
+                abort();
+            }
+            std::string annotationText = "__deegen_asm_annotation_indirectbr{";
+            for (MachineBasicBlock* possibleDestMBB : jumpTables[jumpTableIndex].MBBs)
+            {
+                annotationText += possibleDestMBB->getSymbol()->getName().str() + ",";
+            }
+            annotationText += "}";
+            OutStreamer->AddComment(annotationText);
+          }
+        }
+      }
+
       switch (MI.getOpcode()) {
       case TargetOpcode::CFI_INSTRUCTION:
         emitCFIInstruction(MI);
diff --git a/llvm/lib/IR/ConstantFold.cpp b/llvm/lib/IR/ConstantFold.cpp
index 98adff107..76d408a2a 100644
--- a/llvm/lib/IR/ConstantFold.cpp
+++ b/llvm/lib/IR/ConstantFold.cpp
@@ -30,6 +30,7 @@
 #include "llvm/IR/Operator.h"
 #include "llvm/IR/PatternMatch.h"
 #include "llvm/Support/ErrorHandling.h"
+#include "llvm/IR/ConstantRange.h"
 using namespace llvm;
 using namespace llvm::PatternMatch;

@@ -1387,9 +1388,18 @@ static ICmpInst::Predicate areGlobalsPotentiallyEqual(const GlobalValue *GV1,
     return false;
   };
   // Don't try to decide equality of aliases.
-  if (!isa<GlobalAlias>(GV1) && !isa<GlobalAlias>(GV2))
-    if (!isGlobalUnsafeForEquality(GV1) && !isGlobalUnsafeForEquality(GV2))
-      return ICmpInst::ICMP_NE;
+  if (!isa<GlobalAlias>(GV1) && !isa<GlobalAlias>(GV2)) {
+    if (!isGlobalUnsafeForEquality(GV1) && !isGlobalUnsafeForEquality(GV2)) {
+      if (GV1->isAbsoluteSymbolRef() && GV1->isAbsoluteSymbolRef()) {
+        if (GV1->getAbsoluteSymbolRange()
+             ->intersectWith(GV2->getAbsoluteSymbolRange().value())
+             .isEmptySet())
+          return llvm::CmpInst::ICMP_NE;
+      } else {
+        return ICmpInst::ICMP_NE;
+        }
+      }
+    }
   return ICmpInst::BAD_ICMP_PREDICATE;
 }

@@ -1456,6 +1466,8 @@ static ICmpInst::Predicate evaluateICmpRelation(Constant *V1, Constant *V2,
     if (const GlobalValue *GV2 = dyn_cast<GlobalValue>(V2)) {
       return areGlobalsPotentiallyEqual(GV, GV2);
     } else if (isa<BlockAddress>(V2)) {
+      if (GV->isAbsoluteSymbolRef())
+        return ICmpInst::BAD_ICMP_PREDICATE;
       return ICmpInst::ICMP_NE; // Globals never equal labels.
     } else {
       assert(isa<ConstantPointerNull>(V2) && "Canonicalization guarantee!");
@@ -1466,7 +1478,8 @@ static ICmpInst::Predicate evaluateICmpRelation(Constant *V1, Constant *V2,
       // query it from the Constant type.
       if (!GV->hasExternalWeakLinkage() && !isa<GlobalAlias>(GV) &&
           !NullPointerIsDefined(nullptr /* F */,
-                                GV->getType()->getAddressSpace()))
+                                GV->getType()->getAddressSpace()) &&
+          !GV->isAbsoluteSymbolRef())
         return ICmpInst::ICMP_UGT;
     }
   } else if (const BlockAddress *BA = dyn_cast<BlockAddress>(V1)) {
diff --git a/llvm/lib/Target/AArch64/AArch64CallingConvention.td b/llvm/lib/Target/AArch64/AArch64CallingConvention.td
index c0da242a2..a1fd5c85e 100644
--- a/llvm/lib/Target/AArch64/AArch64CallingConvention.td
+++ b/llvm/lib/Target/AArch64/AArch64CallingConvention.td
@@ -334,9 +334,9 @@ def CC_AArch64_GHC : CallingConv<[
   CCIfType<[v1i64, v2i32, v4i16, v8i8, v2f32], CCBitConvertToType<f64>>,
   CCIfType<[v2i64, v4i32, v8i16, v16i8, v4f32, f128], CCBitConvertToType<v2f64>>,

-  CCIfType<[v2f64], CCAssignToReg<[Q4, Q5]>>,
-  CCIfType<[f32], CCAssignToReg<[S8, S9, S10, S11]>>,
-  CCIfType<[f64], CCAssignToReg<[D12, D13, D14, D15]>>,
+  CCIfType<[v2f64], CCAssignToReg<[Q4, Q5, Q6, Q7, Q8, Q9, Q10]>>,
+  CCIfType<[f32],   CCAssignToReg<[S4, S5, S6, S7, S8, S9, S10]>>,
+  CCIfType<[f64],   CCAssignToReg<[D4, D5, D6, D7, D8, D9, D10]>>,

   // Promote i8/i16/i32 arguments to i64.
   CCIfType<[i8, i16, i32], CCPromoteToType<i64>>,
@@ -424,7 +424,7 @@ def CSR_AArch64_AllRegs
                            (sequence "S%u", 0, 31), (sequence "D%u", 0, 31),
                            (sequence "Q%u", 0, 31))>;

-def CSR_AArch64_NoRegs : CalleeSavedRegs<(add)>;
+def CSR_AArch64_NoRegs : CalleeSavedRegs<(add FP, LR)>;

 def CSR_AArch64_RT_MostRegs :  CalleeSavedRegs<(add CSR_AArch64_AAPCS,
                                                 (sequence "X%u", 9, 15))>;
diff --git a/llvm/lib/Target/AArch64/AArch64FrameLowering.cpp b/llvm/lib/Target/AArch64/AArch64FrameLowering.cpp
index 78babdf9f..356c6de84 100644
--- a/llvm/lib/Target/AArch64/AArch64FrameLowering.cpp
+++ b/llvm/lib/Target/AArch64/AArch64FrameLowering.cpp
@@ -288,7 +288,7 @@ static int64_t getArgumentStackToRestore(MachineFunction &MF,
   } else {
     // ... otherwise the amount to pop is *all* of the argument space,
     // conveniently stored in the MachineFunctionInfo by
-    // LowerFormalArguments. This will, of course, be zero for the C calling
+    // . This will, of course, be zero for the C calling
     // convention.
     ArgumentPopSize = AFI->getArgumentStackToRestore();
   }
@@ -1467,10 +1467,6 @@ void AArch64FrameLowering::emitPrologue(MachineFunction &MF,
     }
   }

-  // All calls are tail calls in GHC calling conv, and functions have no
-  // prologue/epilogue.
-  if (MF.getFunction().getCallingConv() == CallingConv::GHC)
-    return;

   // Set tagged base pointer to the requested stack slot.
   // Ideally it should match SP value after prologue.
@@ -1923,10 +1919,6 @@ void AArch64FrameLowering::emitEpilogue(MachineFunction &MF,
                                : MFI.getStackSize();
   AArch64FunctionInfo *AFI = MF.getInfo<AArch64FunctionInfo>();

-  // All calls are tail calls in GHC calling conv, and functions have no
-  // prologue/epilogue.
-  if (MF.getFunction().getCallingConv() == CallingConv::GHC)
-    return;

   // How much of the stack used by incoming arguments this function is expected
   // to restore in this particular epilogue.
@@ -2948,10 +2940,6 @@ bool AArch64FrameLowering::restoreCalleeSavedRegisters(
 void AArch64FrameLowering::determineCalleeSaves(MachineFunction &MF,
                                                 BitVector &SavedRegs,
                                                 RegScavenger *RS) const {
-  // All calls are tail calls in GHC calling conv, and functions have no
-  // prologue/epilogue.
-  if (MF.getFunction().getCallingConv() == CallingConv::GHC)
-    return;

   TargetFrameLowering::determineCalleeSaves(MF, SavedRegs, RS);
   const AArch64RegisterInfo *RegInfo = static_cast<const AArch64RegisterInfo *>(
diff --git a/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp b/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
index 06e21f90e..84cf49af4 100644
--- a/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
+++ b/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
@@ -6226,6 +6226,7 @@ static bool mayTailCallThisCC(CallingConv::ID CC) {
   case CallingConv::SwiftTail:
   case CallingConv::Tail:
   case CallingConv::Fast:
+  case CallingConv::GHC:
     return true;
   default:
     return false;
@@ -6847,6 +6848,8 @@ AArch64TargetLowering::LowerCall(CallLoweringInfo &CLI,
     if (OpFlags & AArch64II::MO_GOT) {
       Callee = DAG.getTargetGlobalAddress(GV, DL, PtrVT, 0, OpFlags);
       Callee = DAG.getNode(AArch64ISD::LOADgot, DL, PtrVT, Callee);
+    } else if (getTargetMachine().getCodeModel() == CodeModel::Large) {
+      Callee = getAddrLarge(G, DAG, 0);
     } else {
       const GlobalValue *GV = G->getGlobal();
       Callee = DAG.getTargetGlobalAddress(GV, DL, PtrVT, 0, 0);
@@ -6857,7 +6860,9 @@ AArch64TargetLowering::LowerCall(CallLoweringInfo &CLI,
       const char *Sym = S->getSymbol();
       Callee = DAG.getTargetExternalSymbol(Sym, PtrVT, AArch64II::MO_GOT);
       Callee = DAG.getNode(AArch64ISD::LOADgot, DL, PtrVT, Callee);
-    } else {
+    } else if (getTargetMachine().getCodeModel() == CodeModel::Large) {
+      Callee = getAddrLarge(G, DAG, 0);
+    }else {
       const char *Sym = S->getSymbol();
       Callee = DAG.getTargetExternalSymbol(Sym, PtrVT, 0);
     }
@@ -7142,6 +7147,47 @@ SDValue AArch64TargetLowering::getAddrLarge(NodeTy *N, SelectionDAG &DAG,
       getTargetNode(N, Ty, DAG, AArch64II::MO_G0 | MO_NC | Flags));
 }

+template <class NodeTy>
+SDValue AArch64TargetLowering::getAddrAbsolute(const GlobalValue* GV, NodeTy *N,
+                                               SelectionDAG &DAG,
+                                               unsigned Flags) const {
+  LLVM_DEBUG(dbgs() << "AArch64TargetLowering::getAddrAbsolute\n");
+  SDLoc DL(N);
+  EVT Ty = getPointerTy(DAG.getDataLayout());
+  ConstantRange CR = GV->getAbsoluteSymbolRange().value();
+  unsigned Width = CR.getUnsignedMax().getActiveBits();
+
+  SDValue Cur;
+  bool First = true;
+
+  auto BuildMov = [&](unsigned MovFlags, unsigned Shift) {
+    SDValue Granule = DAG.getTargetGlobalAddress(
+        GV, DL, Ty, 0, MovFlags | (First ? 0 : AArch64II::MO_NC) | Flags);
+    if (First) {
+      First = false;
+      return SDValue(DAG.getMachineNode(AArch64::MOVZXi, DL, Ty, Granule,
+                                        DAG.getTargetConstant(Shift, DL, MVT::i32)),
+                                        0);
+    } else {
+      return SDValue(DAG.getMachineNode(AArch64::MOVKXi, DL, Ty, Cur, Granule,
+                                        DAG.getTargetConstant(Shift, DL, MVT::i32)),
+                                         0);
+    }
+  };
+
+  if (48 < Width) {
+    Cur = BuildMov(AArch64II::MO_G3, 48);
+  }
+  if (32 < Width) {
+    Cur = BuildMov(AArch64II::MO_G2, 32);
+  }
+  if (16 < Width) {
+    Cur = BuildMov(AArch64II::MO_G1, 16);
+  }
+  return BuildMov(AArch64II::MO_G0, 0);
+}
+
+
 // (addlow (adrp %hi(sym)) %lo(sym))
 template <class NodeTy>
 SDValue AArch64TargetLowering::getAddr(NodeTy *N, SelectionDAG &DAG,
@@ -7184,7 +7230,9 @@ SDValue AArch64TargetLowering::LowerGlobalAddress(SDValue Op,
   }

   SDValue Result;
-  if (getTargetMachine().getCodeModel() == CodeModel::Large) {
+  if (GV->isAbsoluteSymbolRef()) {
+    Result = getAddrAbsolute(GV, GN, DAG, OpFlags);
+  } else if (getTargetMachine().getCodeModel() == CodeModel::Large) {
     Result = getAddrLarge(GN, DAG, OpFlags);
   } else if (getTargetMachine().getCodeModel() == CodeModel::Tiny) {
     Result = getAddrTiny(GN, DAG, OpFlags);
diff --git a/llvm/lib/Target/AArch64/AArch64ISelLowering.h b/llvm/lib/Target/AArch64/AArch64ISelLowering.h
index ff3bfe897..3ec69a193 100644
--- a/llvm/lib/Target/AArch64/AArch64ISelLowering.h
+++ b/llvm/lib/Target/AArch64/AArch64ISelLowering.h
@@ -949,6 +949,8 @@ private:
   template <class NodeTy>
   SDValue getAddrLarge(NodeTy *N, SelectionDAG &DAG, unsigned Flags = 0) const;
   template <class NodeTy>
+  SDValue getAddrAbsolute(const GlobalValue* GV, NodeTy *N, SelectionDAG &DAG, unsigned Flags = 0) const;
+  template <class NodeTy>
   SDValue getAddr(NodeTy *N, SelectionDAG &DAG, unsigned Flags = 0) const;
   template <class NodeTy>
   SDValue getAddrTiny(NodeTy *N, SelectionDAG &DAG, unsigned Flags = 0) const;
diff --git a/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp b/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp
index 835a7b6cc..8c18380b4 100644
--- a/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp
+++ b/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp
@@ -2556,6 +2556,48 @@ bool AArch64InstrInfo::getMemOperandsWithOffsetWidth(
   return true;
 }

+int AArch64InstrInfo::getJumpTableIndex(const llvm::MachineInstr &MI) const {
+  unsigned Opcode = MI.getOpcode();
+  // $x10, $x11 = JumpTableDest8 $x9, $x8, %XX
+  // BR $x10
+  if (Opcode == AArch64::BR) {
+    Register Reg = MI.getOperand(0).getReg();
+    const MachineInstr *JTD;
+
+    if (Reg.isVirtual()) {
+      const MachineFunction &MF = *MI.getParent()->getParent();
+      const MachineRegisterInfo &MRI = MF.getRegInfo();
+      JTD = MRI.getUniqueVRegDef(Reg);
+    } else {
+      const MachineBasicBlock *MBB = MI.getParent();
+      if (!MBB)
+        return -1;
+      auto I = MI.getReverseIterator();
+      for (; I != MBB->rend(); ++I) {
+        for (const MachineOperand &MO : I->defs()) {
+          if (MO.isReg() && MO.getReg() == Reg && !MO.isImplicit()) {
+            JTD = &*I;
+          }
+        }
+      }
+    }
+
+    if (JTD == nullptr)
+      return -1;
+
+    if (JTD->getOpcode() != AArch64::JumpTableDest8 &&
+        JTD->getOpcode() != AArch64::JumpTableDest16 &&
+        JTD->getOpcode() != AArch64::JumpTableDest32)
+      return -1;
+
+    if (!JTD->getOperand(4).isJTI())
+      return -1;
+
+    return JTD->getOperand(4).getIndex();
+  }
+  return -1;
+}
+
 Optional<ExtAddrMode>
 AArch64InstrInfo::getAddrModeFromMemoryOp(const MachineInstr &MemI,
                                           const TargetRegisterInfo *TRI) const {
diff --git a/llvm/lib/Target/AArch64/AArch64InstrInfo.h b/llvm/lib/Target/AArch64/AArch64InstrInfo.h
index b7a6ac301..ece4f3a49 100644
--- a/llvm/lib/Target/AArch64/AArch64InstrInfo.h
+++ b/llvm/lib/Target/AArch64/AArch64InstrInfo.h
@@ -134,6 +134,8 @@ public:
   /// Hint that pairing the given load or store is unprofitable.
   static void suppressLdStPair(MachineInstr &MI);

+  int getJumpTableIndex(const MachineInstr &MI) const override;
+
   Optional<ExtAddrMode>
   getAddrModeFromMemoryOp(const MachineInstr &MemI,
                           const TargetRegisterInfo *TRI) const override;
diff --git a/llvm/lib/Target/AArch64/AArch64Subtarget.cpp b/llvm/lib/Target/AArch64/AArch64Subtarget.cpp
index 150053043..31cfedb33 100644
--- a/llvm/lib/Target/AArch64/AArch64Subtarget.cpp
+++ b/llvm/lib/Target/AArch64/AArch64Subtarget.cpp
@@ -314,6 +314,10 @@ const RegisterBankInfo *AArch64Subtarget::getRegBankInfo() const {
 unsigned
 AArch64Subtarget::ClassifyGlobalReference(const GlobalValue *GV,
                                           const TargetMachine &TM) const {
+  // Absolute symbols are always MO_NO_FLAG to enable proper materialization
+  if (GV->isAbsoluteSymbolRef())
+    return AArch64II::MO_NO_FLAG;
+
   // MachO large model always goes via a GOT, simply to get a single 8-byte
   // absolute relocation on all global addresses.
   if (TM.getCodeModel() == CodeModel::Large && isTargetMachO())
diff --git a/llvm/lib/Target/X86/X86CallingConv.td b/llvm/lib/Target/X86/X86CallingConv.td
index 4dd8a6cdd..5a2c1fa5d 100644
--- a/llvm/lib/Target/X86/X86CallingConv.td
+++ b/llvm/lib/Target/X86/X86CallingConv.td
@@ -705,9 +705,9 @@ def CC_X86_64_GHC : CallingConv<[
   // Promote i8/i16/i32 arguments to i64.
   CCIfType<[i8, i16, i32], CCPromoteToType<i64>>,

-  // Pass in STG registers: Base, Sp, Hp, R1, R2, R3, R4, R5, R6, SpLim
+  // Pass in STG registers: Base, Sp, Hp, R1, R2, R3, R4, R5, R6, SpLim, RCX, RDX, R10, R11
   CCIfType<[i64],
-            CCAssignToReg<[R13, RBP, R12, RBX, R14, RSI, RDI, R8, R9, R15]>>,
+            CCAssignToReg<[R13, RBP, R12, RBX, R14, RSI, RDI, R8, R9, R15, RCX, RDX, R10, R11]>>,

   // Pass in STG registers: F1, F2, F3, F4, D1, D2
   CCIfType<[f32, f64, v16i8, v8i16, v4i32, v2i64, v4f32, v2f64],
@@ -1154,14 +1154,14 @@ def CSR_64_CXX_TLS_Darwin_PE : CalleeSavedRegs<(add RBP)>;
 // CSRs that are handled explicitly via copies.
 def CSR_64_CXX_TLS_Darwin_ViaCopy : CalleeSavedRegs<(sub CSR_64_TLS_Darwin, RBP)>;

-// All GPRs - except r11
-def CSR_64_RT_MostRegs : CalleeSavedRegs<(add CSR_64, RAX, RCX, RDX, RSI, RDI,
+// All GPRs - except r11, rax and rdx
+def CSR_64_RT_MostRegs : CalleeSavedRegs<(add CSR_64, RCX, RSI, RDI,
                                               R8, R9, R10)>;

 // All registers - except r11
-def CSR_64_RT_AllRegs     : CalleeSavedRegs<(add CSR_64_RT_MostRegs,
+def CSR_64_RT_AllRegs     : CalleeSavedRegs<(add CSR_64_RT_MostRegs, RAX, RDX,
                                                  (sequence "XMM%u", 0, 15))>;
-def CSR_64_RT_AllRegs_AVX : CalleeSavedRegs<(add CSR_64_RT_MostRegs,
+def CSR_64_RT_AllRegs_AVX : CalleeSavedRegs<(add CSR_64_RT_MostRegs, RAX, RDX,
                                                  (sequence "YMM%u", 0, 15))>;

 def CSR_64_MostRegs : CalleeSavedRegs<(add RBX, RCX, RDX, RSI, RDI, R8, R9, R10,
diff --git a/llvm/lib/Target/X86/X86ISelDAGToDAG.cpp b/llvm/lib/Target/X86/X86ISelDAGToDAG.cpp
index f88037e95..63eeebd19 100644
--- a/llvm/lib/Target/X86/X86ISelDAGToDAG.cpp
+++ b/llvm/lib/Target/X86/X86ISelDAGToDAG.cpp
@@ -1786,6 +1786,14 @@ bool X86DAGToDAGISel::matchWrapper(SDValue N, X86ISelAddressMode &AM) {
     return true;
   }

+  if (Subtarget->is64Bit() && AM.GV && AM.GV->isAbsoluteSymbolRef()) {
+  Optional<ConstantRange> CR = AM.GV->getAbsoluteSymbolRange();
+    if (CR.getValue().getMinSignedBits() > 32) {
+      AM = Backup;
+      return true;
+    }
+  }
+
   if (IsRIPRel)
     AM.setBaseReg(CurDAG->getRegister(X86::RIP, MVT::i64));

@@ -2693,7 +2701,7 @@ bool X86DAGToDAGISel::selectMOV64Imm32(SDValue N, SDValue &Imm) {
   if (!CR)
     return TM.getCodeModel() == CodeModel::Small;

-  return CR->getUnsignedMax().ult(1ull << 32);
+  return CR->getMinSignedBits() <= 32;
 }

 bool X86DAGToDAGISel::selectLEA64_32Addr(SDValue N, SDValue &Base,
diff --git a/llvm/lib/Target/X86/X86InstrInfo.cpp b/llvm/lib/Target/X86/X86InstrInfo.cpp
index 74ef831e1..613faece6 100644
--- a/llvm/lib/Target/X86/X86InstrInfo.cpp
+++ b/llvm/lib/Target/X86/X86InstrInfo.cpp
@@ -3252,6 +3252,68 @@ bool X86InstrInfo::analyzeBranchPredicate(MachineBasicBlock &MBB,
   return true;
 }

+static int getJumpTableIndexFromAddr(const MachineInstr &MI) {
+    const MCInstrDesc &Desc = MI.getDesc();
+    int MemRefBegin = X86II::getMemoryOperandNo(Desc.TSFlags);
+    assert(MemRefBegin >= 0 && "instr should have memory operand");
+    MemRefBegin += X86II::getOperandBias(Desc);
+
+    const MachineOperand &MO = MI.getOperand(MemRefBegin + X86::AddrDisp);
+    if (!MO.isJTI())
+        return -1;
+
+    return MO.getIndex();
+}
+
+static int getJumpTableIndexFromReg(const MachineRegisterInfo &MRI,
+                                    Register Reg) {
+    if (!Reg.isVirtual())
+        return -1;
+    MachineInstr *MI = MRI.getUniqueVRegDef(Reg);
+    if (MI == nullptr)
+        return -1;
+    unsigned Opcode = MI->getOpcode();
+    if (Opcode != X86::LEA64r && Opcode != X86::LEA32r)
+        return -1;
+    return getJumpTableIndexFromAddr(*MI);
+}
+
+int X86InstrInfo::getJumpTableIndex(const MachineInstr &MI) const {
+    unsigned Opcode = MI.getOpcode();
+    // Switch-jump pattern for non-PIC code looks like:
+    //   JMP64m $noreg, 8, %X, %jump-table.X, $noreg
+    if (Opcode == X86::JMP64m || Opcode == X86::JMP32m) {
+        return getJumpTableIndexFromAddr(MI);
+    }
+    // The pattern for PIC code looks like:
+    //   %0 = LEA64r $rip, 1, $noreg, %jump-table.X
+    //   %1 = MOVSX64rm32 %0, 4, XX, 0, $noreg
+    //   %2 = ADD64rr %1, %0
+    //   JMP64r %2
+    if (Opcode == X86::JMP64r || Opcode == X86::JMP32r) {
+        Register Reg = MI.getOperand(0).getReg();
+        if (!Reg.isVirtual())
+            return -1;
+
+        const MachineFunction &MF = *MI.getParent()->getParent();
+        const MachineRegisterInfo &MRI = MF.getRegInfo();
+        MachineInstr *Add = MRI.getUniqueVRegDef(Reg);
+        if (Add == nullptr)
+            return -1;
+
+        if (Add->getOpcode() != X86::ADD64rr && Add->getOpcode() != X86::ADD32rr)
+            return -1;
+
+        int JTI1 = getJumpTableIndexFromReg(MRI, Add->getOperand(1).getReg());
+        if (JTI1 >= 0)
+            return JTI1;
+        int JTI2 = getJumpTableIndexFromReg(MRI, Add->getOperand(2).getReg());
+        if (JTI2 >= 0)
+            return JTI2;
+    }
+    return -1;
+}
+
 unsigned X86InstrInfo::removeBranch(MachineBasicBlock &MBB,
                                     int *BytesRemoved) const {
   assert(!BytesRemoved && "code size not handled");
diff --git a/llvm/lib/Target/X86/X86InstrInfo.h b/llvm/lib/Target/X86/X86InstrInfo.h
index 81729e361..abc387bcd 100644
--- a/llvm/lib/Target/X86/X86InstrInfo.h
+++ b/llvm/lib/Target/X86/X86InstrInfo.h
@@ -327,6 +327,8 @@ public:
                      SmallVectorImpl<MachineOperand> &Cond,
                      bool AllowModify) const override;

+  int getJumpTableIndex(const MachineInstr &MI) const override;
+
   Optional<ExtAddrMode>
   getAddrModeFromMemoryOp(const MachineInstr &MemI,
                           const TargetRegisterInfo *TRI) const override;
